File: c_sharp/Scope.cs
using System;
using System.Collections.Generic;
using System.Text;
using whisper_language.gen;

namespace whisper_language
{
    class Scope
    {
        private Scope parent;
        
        Dictionary<String, TLValue> variables;
        
        private bool isFunction;

        internal Scope Parent { get => parent; set => parent = value; }

        public Scope(Scope p, bool function)
        {
            Parent = p;
            variables = new Dictionary<string, TLValue>();
            isFunction = function;
        }

        public void assignParam(String var, TLValue value)
        {
            variables.Add(var, value);
        }

        public void assign(String var, TLValue value)
        {
            if (resolve(var, !isFunction) != null)
            {
                // There is already such a variable, re-assign it
                this.reAssign(var, value);
            }
            else
            {
                // A newly declared variable
                variables.Add(var, value);
            }
        }

        private bool isGlobalScope()
        {
            return Parent == null;
        }

        private void reAssign(String identifier, TLValue value)
        {
            if (variables.ContainsKey(identifier))
            {
                // The variable is declared in this scope
                variables.Remove(identifier);
                variables.Add(identifier, value);
            }
            else if (Parent != null)
            {
                // The variable was not declared in this scope, so let
                // the parent scope re-assign it
                Parent.reAssign(identifier, value);
            }
        }

        public TLValue resolve(String var)
        {
            return resolve(var, true);
        }

        private TLValue resolve(String var, bool checkParent)
        { 
            TLValue value = variables.ContainsKey(var) ? variables[var]:null;
            if (value != null)
            {
                // The variable resides in this scope
                return value;
            }
            else if (checkParent && !isGlobalScope())
            {
                // Let the parent scope look for the variable
                return Parent.resolve(var, !Parent.isFunction);
            }
            else
            {
                // Unknown variable
                return null;
            }
        }

    }
}

File: dart/bin/Scope.dart
import 'dart:core';

import 'TLValue.dart';

class Scope {
  Scope parent;

  Map variables = {};

  bool isFunction = false;

  Scope(this.parent, this.isFunction);

  //
  void assignParam(String var1, TLValue value) {
    variables[var1] = value;
  }

  void assign(String var1, TLValue value) {
    if (resolve(var1, !isFunction) != null) {
      reAssign(var1, value);
    } else {
      // A newly declared variable
      variables[var1] = value;
    }
  }

  void reAssign(String identifier, TLValue value) {
    if (variables.containsKey(identifier)) {
      // The variable is declared in this scope
      // print("write current scope "+identifier);
      variables[identifier] = value;
    } else if (parent != null) {
      // print("write parent scope "+identifier);
      // The variable was not declared in this scope, so let
      // the parent scope re-assign it
      parent.reAssign(identifier, value);
    } else {
      print("parent is null!!");
    }
  }

  bool isGlobalScope() {
    return parent == null;
  }

  bool isFunc() {
    return isFunction = true;
  }

  TLValue resolve(String var1, bool checkParent) {
    TLValue value = variables[var1];
    // print(parent);
    // print("${var1} ${checkParent} ${!isGlobalScope()}");
    if (value != null) {
      return value;
    } else if (checkParent && !isGlobalScope()) {
      // Let the parent scope look for the variable
      return parent.resolve(var1, !(parent.isFunction));
    } else {
      print("没有找到");
      return null;
    }
  }
}
